miner/src/block_assembler.rs

```diff
@@ -413,6 +418,192 @@ impl BlockAssembler {
         Ok(template)
     }
 
+    fn package_txs(
+        &self,
+        chain_state: &ChainState,
+        size_limit: usize,
+        cycles_limit: Cycle,
+    ) -> (Vec<TxEntry>, usize, Cycle) {
+        /// update descendants txs into modified entries
+        fn update_modified_entries(
+            tx_pool: &TxPool,
+            in_block_txs: &HashSet<ProposalShortId>,
+            package_txs: &HashSet<TxEntry>,
+            modified_entries: &mut TxModifiedEntries,
+        ) {
+            for ptx in package_txs {
+                let ptx_id = ptx.transaction.proposal_short_id();
+                if in_block_txs.contains(&ptx_id) {
+                    continue;
+                }
+                let descendants = tx_pool.proposed().get_descendants(&ptx_id);
+                for id in descendants {
+                    let mut tx = match modified_entries.remove(&id) {
+                        Some(modified_tx) => modified_tx,
+                        None => tx_pool
+                            .proposed()
+                            .get(&id)
+                            .map(ToOwned::to_owned)
+                            .expect("pool consistent"),
+                    };
+                    tx.sub_entry_weight(&ptx);
+                    modified_entries.insert(tx);
+                }
+            }
+        }
+
+        /// find a valid tx_entry
+        fn find_tx_entry<F: Fn(&TxEntry) -> bool>(
```

Comment:
The code has many duplicated logic with `std::iter::Filter`，or `std::iter::Filter` + `std::iter::TakeWhilte`.

Reply:
We need to maintain a failure counter in the loop, use iterator is hard to achieve this.